src/storage/index/IndexExecutor.inl

```diff
@@ -0,0 +1,329 @@
+            const auto& ftype = schema->getFieldType(col);
+            if (UNLIKELY(ftype == CommonConstants::kInvalidValueType())) {
+                return cpp2::ErrorCode::E_IMPROPER_DATA_TYPE;
+            }
+            /**
+             * Create PropContext for getRowFromReader
+             */
+            {
+                PropContext prop;
+                prop.type_ = ftype;
+                prop.retIndex_ = index++;
+                storage::cpp2::PropDef pd;
+                pd.set_name(col);
+                prop.prop_ = std::move(pd);
+                prop.returned_ = true;
+                props_.emplace_back(std::move(prop));
+            }
+            schema_->appendCol(col, std::move(ftype).get_type());
+        }   // end for
+    }
+    return cpp2::ErrorCode::SUCCEEDED;
+}
+
+template <typename RESP>
+kvstore::ResultCode IndexExecutor<RESP>::executeExecutionPlan(PartitionID part) {
+    std::string prefix = NebulaKeyUtils::indexPrefix(part, index_->get_index_id())
+                        .append(prefix_);
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto ret = this->kvstore_->prefix(spaceId_,
+                                      part,
+                                      prefix,
+                                      &iter);
+    if (ret != nebula::kvstore::SUCCEEDED) {
+        return ret;
+    }
+    while (iter->valid() &&
+           rowNum_ < FLAGS_max_rows_returned_per_lookup) {
+        auto key = iter->key();
+        /**
+         * Need to filter result with expression if is not accurate scan.
+         */
+        if (isIndexScan_ && !conditionsCheck(key)) {
+            iter->next();
+            continue;
+        }
+        ret = getDataRow(part, key);
+        if (ret != kvstore::ResultCode::SUCCEEDED) {
+            return ret;
+        }
+        iter->next();
+    }
+    return ret;
+}
+
+template<typename RESP>
+kvstore::ResultCode IndexExecutor<RESP>::getDataRow(PartitionID partId,
+                                                    const folly::StringPiece& key) {
+    kvstore::ResultCode ret;
+    if (isEdgeIndex_) {
+        cpp2::Edge data;
+        ret = getEdgeRow(partId, key, &data);
+        if (ret == kvstore::SUCCEEDED) {
+            edgeRows_.emplace_back(std::move(data));
+            ++rowNum_;
+        }
+    } else {
+        cpp2::VertexIndexData data;
+        ret = getVertexRow(partId, key, &data);
+        if (ret == kvstore::SUCCEEDED) {
+            vertexRows_.emplace_back(std::move(data));
+            ++rowNum_;
+        }
+    }
+    return ret;
+}
+
+template<typename RESP>
+kvstore::ResultCode IndexExecutor<RESP>::getVertexRow(PartitionID partId,
+                                                      const folly::StringPiece& key,
+                                                      cpp2::VertexIndexData* data) {
+    auto vId = NebulaKeyUtils::getIndexVertexID(key);
+    data->set_vertex_id(vId);
+    if (schema_ == nullptr) {
```

Comment:
`schema_ == null` Is that possible?

Reply:
> `schema_ == null` Is that possible?

Emm… the schema_ is a shared_ptr as below : 
std::shared_ptr<SchemaWriter>          schema_{nullptr};

Reply:
when `checkReturnColumns` the schema have constructed.

BTW is the schema is null why return `SUCCEEDED`?

Reply:
> when `checkReturnColumns` the schema have constructed.
> 
> BTW is the schema is null why return `SUCCEEDED`?

Aha, The result set such as `tag_index(VertexID , col1, col2...) `or `edge_index(srcVID, dstVID, Ranking, col1 , col2...)` , if the yield clause is empty , so we only need return (VertexID) or (srcVID, dstVID, Ranking) . So we don't have to handle with the schema, so return SUCCEEDED direct.